hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableImpl.java

```diff
@@ -0,0 +1,271 @@
+        if (listener == null) {
+            throw new IllegalArgumentException(
+                    String.format("No registered observer with registration ID %s", registrationId));
+        } else {
+            listener.cancel();
+        }
+    }
+
+    @Override
+    public void destroy() {
+        listeners.keySet().forEach(this::removeObserver);
+        listeners.clear();
+
+        onDestroy.accept(this);
+    }
+
+    private static class RingbufferListener<T> {
+
+        private static final int BATCH_SIZE = RingbufferProxy.MAX_BATCH_SIZE;
+
+        private final String observableName;
+        private final Observer<T> observer;
+
+        private final Ringbuffer<Object> ringbuffer;
+        private final ILogger logger;
+        private final Executor executor;
+
+        private long sequence;
+        private volatile boolean cancelled;
+
+        RingbufferListener(
+                String observableName,
+                Observer<T> observer,
+                HazelcastInstance hzInstance,
+                ILogger logger
+        ) {
+            this.observableName = observableName;
+            this.observer = observer;
+            this.ringbuffer = hzInstance.getRingbuffer(ObservableUtil.getRingbufferName(observableName));
+            this.logger = logger;
+            this.executor = getExecutor(hzInstance);
+
+            this.sequence = ringbuffer.headSequence();
+        }
+
+        void next() {
+            if (cancelled) {
+                return;
+            }
+            ringbuffer.readManyAsync(sequence, 1, BATCH_SIZE, null)
+                    .whenCompleteAsync(this::accept, executor);
+        }
+
+        void cancel() {
+            cancelled = true;
+        }
+
+        private void accept(ReadResultSet<Object> resultSet, Throwable throwable) {
+            if (cancelled) {
+                return;
+            }
+            if (throwable == null) {
+                // we process all messages in batch. So we don't release the thread and reschedule ourselves;
+                // but we'll process whatever was received in 1 go.
+                for (Object result : resultSet) {
+                    try {
+                        onNewMessage(result);
+                    } catch (Throwable t) {
+                        logger.warning("Terminating message listener on ring-buffer " + ringbuffer.getName() +
+                                ". Reason: Unhandled exception, message: " + t.getMessage(), t);
+                        cancel();
+                        return;
+                    }
+
+                    sequence++;
+                }
+                next();
+            } else {
+                if (handleInternalException(throwable)) {
+                    next();
+                } else {
+                    cancel();
+                }
+            }
+        }
+
+        private void onNewMessage(Object message) {
+            if (message instanceof Throwable) {
```

Comment:
This will clash if for whatever reason the user sends `Throwable`s to the sink :). Maybe it's not worth fixing it, bu twe can mention in the `onError` method javadoc that it's called also if a `Throwable` is added to the sink...